File: crates/treeman-core/src/slug.rs
//! Slug derivation. Ports the bash `slug_for()` from `.gwt-postcreate` /
//! `.gwt-predelete` so a worktree that was named by `gwt` keeps the same
//! identifier when treeman owns the lifecycle.
//!
//! Rules:
//!   - If a branch name (or, as fallback, the last path component) contains
//!     `[A-Z]+-\d+` (a Jira ticket key), the slug is `<prefix>_<number>`
//!     lowercased.
//!   - Otherwise, slug = `wt_<blake3(canonical-path)[..8]>` — deterministic
//!     and distinct between two worktrees whose basenames collide.
//!
//! Derived properties (`slug_dash`, `slug_redis_queue`, `slug_redis_cache`)
//! match the bash impl exactly — see `derive_redis_indices` for the cksum
//! quirk: bash `cksum` is the SysV checksum, which we re-implement here so
//! a worktree that already exists doesn't change its redis db on the switch
//! to treeman.

use std::path::Path;

use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Slug {
    pub value: String,
    pub source: SlugSource,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SlugSource {
    Ticket,
    PathHash,
}

impl Slug {
    pub fn as_str(&self) -> &str {
        &self.value
    }
    pub fn dashed(&self) -> String {
        self.value.replace('_', "-")
    }
    /// (queue_db, cache_db) — bash `cksum` parity. See module-level docs.
    pub fn redis_indices(&self) -> (u8, u8) {
        let h = sysv_cksum(self.value.as_bytes());
        let queue = (h % 10 + 6) as u8;
        let cache = ((h / 10) % 10 + 6) as u8;
        (queue, cache)
    }
}

/// `branch` is consulted first (matches the bash flow: a ticket-named
/// branch wins even if the worktree dir was named generically). Then the
/// worktree's last path component is checked. Finally we hash the
/// canonical path.
pub fn slug_for(worktree_path: &Path, branch: Option<&str>) -> Slug {
    let ticket_re = ticket_regex();
    if let Some(b) = branch {
        if let Some(s) = extract_ticket(b, &ticket_re) {
            return Slug {
                value: s,
                source: SlugSource::Ticket,
            };
        }
    }
    if let Some(last) = worktree_path.file_name().and_then(|s| s.to_str()) {
        if let Some(s) = extract_ticket(last, &ticket_re) {
            return Slug {
                value: s,
                source: SlugSource::Ticket,
            };
        }
    }
    let canonical = worktree_path
        .canonicalize()
        .unwrap_or_else(|_| worktree_path.to_path_buf());
    let bytes = canonical.as_os_str().as_encoded_bytes();
    let h = blake3::hash(bytes);
    let hex = h.to_hex();
    Slug {
        value: format!("wt_{}", &hex[..8]),
        source: SlugSource::PathHash,
    }
}

fn extract_ticket(s: &str, re: &Regex) -> Option<String> {
    re.captures(s).map(|caps| {
        let prefix = caps.get(1).unwrap().as_str().to_lowercase();
        let num = caps.get(2).unwrap().as_str();
        let mut out = format!("{prefix}_{num}");
        if out.len() > 32 {
            out.truncate(32);
        }
        out
    })
}

fn ticket_regex() -> Regex {
    // [A-Z]+-[0-9]+
    Regex::new(r"([A-Z]+)-([0-9]+)").expect("static regex")
}

/// SysV `cksum` (POSIX) — matches `printf '%s' SLUG | cksum | awk '{print $1}'`
/// used by the bash hook. We must reproduce the exact algorithm so a
/// worktree's redis db indices don't shift on the switch from bash → treeman.
fn sysv_cksum(input: &[u8]) -> u32 {
    // POSIX cksum: CRC-32 over (data || length-bytes-little-endian-by-octets).
    // Algorithm per `man cksum`. Initial CRC=0, no inversion, polynomial
    // 0x04C11DB7 with the high bit out. We compute it byte-by-byte; the
    // length suffix encodes each octet of the length, LSB first, ending
    // when the remaining length is zero.
    let mut crc: u32 = 0;
    for &b in input {
        crc = crc_update(crc, b);
    }
    let mut len = input.len() as u64;
    while len != 0 {
        crc = crc_update(crc, (len & 0xff) as u8);
        len >>= 8;
    }
    !crc
}

fn crc_update(mut crc: u32, byte: u8) -> u32 {
    crc ^= (byte as u32) << 24;
    for _ in 0..8 {
        if crc & 0x8000_0000 != 0 {
            crc = (crc << 1) ^ 0x04C1_1DB7;
        } else {
            crc <<= 1;
        }
    }
    crc
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    #[test]
    fn ticket_in_branch_wins() {
        let s = slug_for(
            &PathBuf::from("/tmp/random-dir"),
            Some("feature/KON-1234-foo"),
        );
        assert_eq!(s.value, "kon_1234");
        assert_eq!(s.source, SlugSource::Ticket);
    }

    #[test]
    fn ticket_in_path_basename() {
        let s = slug_for(&PathBuf::from("/tmp/KON-9001-bar"), None);
        assert_eq!(s.value, "kon_9001");
    }

    #[test]
    fn falls_back_to_path_hash() {
        let s = slug_for(&PathBuf::from("/nonexistent/random"), None);
        assert!(s.value.starts_with("wt_"));
        assert_eq!(s.value.len(), 11);
        assert_eq!(s.source, SlugSource::PathHash);
    }

    #[test]
    fn slug_dashed_replaces_underscores() {
        let s = Slug {
            value: "kon_1234".into(),
            source: SlugSource::Ticket,
        };
        assert_eq!(s.dashed(), "kon-1234");
    }

    #[test]
    fn redis_indices_in_range_6_15() {
        let s = Slug {
            value: "kon_1234".into(),
            source: SlugSource::Ticket,
        };
        let (q, c) = s.redis_indices();
        assert!((6..=15).contains(&q));
        assert!((6..=15).contains(&c));
    }

    #[test]
    fn sysv_cksum_known_vectors() {
        // `printf '' | cksum`              => 4294967295 (0xFFFFFFFF), 0
        // `printf 'a' | cksum`             => 1220704766, 1
        // `printf 'abc' | cksum`           => 1219131554, 3
        // `printf 'kon_1234' | cksum`      => verified via `printf 'kon_1234' | cksum` on linux
        assert_eq!(sysv_cksum(b""), 4294967295);
        assert_eq!(sysv_cksum(b"a"), 1220704766);
        assert_eq!(sysv_cksum(b"abc"), 1219131554);
    }
}

File: internal/slug/slug.go
// Package slug derives a worktree slug from the branch name / path,
// and exposes deterministic redis-db indices so worktrees retain
// their per-engine namespace across treeman upgrades.
//
// The SysV cksum reimplementation matches POSIX `cksum` byte-for-
// byte so external shell hooks that compute the same redis indices
// agree with treeman's slugs.
package slug

import (
	"path/filepath"
	"regexp"
	"strings"

	"lukechampine.com/blake3"
)

// Source records where the slug came from.
type Source int

const (
	// SourceTicket — the slug was extracted from a Jira-style ticket
	// pattern (`[A-Z]+-\d+`) in the branch name or directory.
	SourceTicket Source = iota
	// SourcePathHash — fallback when no ticket was found; slug is a
	// stable `wt_<blake3(path)[:8]>`.
	SourcePathHash
	// SourceMain — derived by ForMain() for a repo's main worktree.
	// Always branch-driven (`main_<sanitised_branch>`) so a non-ticket
	// branch like `develop` doesn't collapse to the repo's path hash,
	// which would share a DB across every branch ever checked out at
	// the repo root.
	SourceMain
)

// Slug is the worktree identifier feeding every per-engine template
// (`{slug}`, `{slug_dash}`, `{slug_redis_queue}`, `{slug_redis_cache}`).
type Slug struct {
	Value  string
	Source Source
}

// Dashed returns the slug with `_` replaced by `-`. Used by the
// `{slug_dash}` template key for S3 / minio buckets where
// underscores are invalid.
func (s Slug) Dashed() string {
	return strings.ReplaceAll(s.Value, "_", "-")
}

// RedisIndices returns (queue_db, cache_db), each in 6..15, derived
// deterministically from the slug via POSIX `cksum`.
func (s Slug) RedisIndices() (queueDB, cacheDB uint8) {
	h := sysvCksum([]byte(s.Value))
	queueDB = uint8(h%10 + 6)
	cacheDB = uint8((h/10)%10 + 6)
	return
}

var ticketRe = regexp.MustCompile(`([A-Z]+)-([0-9]+)`)

// For derives a linked worktree's slug purely from its filesystem
// PATH. The `branch` parameter is accepted for signature compatibility
// with the call sites that happen to know the branch, but is
// deliberately IGNORED — folding the branch into the slug is what the
// two failure modes this function guards against both come down to.
//
// Two properties fall out of keying strictly on the canonical path:
//
//   - Branch-stable. The directory is fixed for a worktree's lifetime,
//     so the slug — and with it the worktree's DB names, patched
//     `.env`, ports, redis indices — does NOT churn when the branch
//     swaps underneath it via an in-worktree `git checkout`. This is
//     the foundation `branch_scoped` swapping relies on, and the reason
//     ResolveIdentity has always called us with an empty branch.
//
//   - Collision-free across a shared ticket. Two worktrees whose names
//     embed the SAME Jira ticket (`PROJ-1234` checked out in two
//     different directories, or two branches of one ticket) get
//     DISTINCT slugs, because the path hash always disambiguates. The
//     previous derivation returned a bare `proj_1234` for both and
//     silently overlapped their storage — same DB names, same redis
//     DBs, same patched `.env`.
//
// A ticket-named directory keeps a readable prefix for ergonomics
// (`proj_1234_<pathhash>`); a generic directory is `wt_<pathhash>`. The
// path hash is the sole carrier of uniqueness in both shapes.
func For(worktreePath string, _ string) Slug {
	tag := pathTag(worktreePath)
	if ticket, ok := extractTicket(filepath.Base(worktreePath)); ok {
		return Slug{Value: composeTicket(ticket, tag), Source: SourceTicket}
	}
	return Slug{Value: "wt_" + tag, Source: SourcePathHash}
}

// pathTag is a stable 8-hex-char blake3 digest of the worktree's
// canonical absolute path — the collision-proof component of every
// linked-worktree slug. Falls back to the raw path when Abs fails
// (e.g. the cwd was removed) so a slug is still produced.
func pathTag(worktreePath string) string {
	canonical, err := filepath.Abs(worktreePath)
	if err != nil {
		canonical = worktreePath
	}
	sum := blake3.Sum256([]byte(canonical))
	return hexEncode(sum[:])[:8]
}

// composeTicket joins a readable ticket prefix with the path tag inside
// the 32-char identifier budget every engine's name template must fit.
// Uniqueness lives in the tag, never the prefix, so truncating an
// over-long prefix can't collide two distinct paths — their tags still
// differ.
func composeTicket(ticket, tag string) string {
	out := ticket + "_" + tag
	if len(out) <= 32 {
		return out
	}
	keep := max(32-1-len(tag), 1)
	return ticket[:keep] + "_" + tag
}

// ForMain derives the slug for a repo's main worktree (the repo root
// checkout). Unlike For(), the branch is mandatory and there's no
// path-hash fallback — checking out a non-ticket branch like `develop`
// at the repo root must produce a branch-specific slug (`main_develop`)
// so per-branch DBs in the main wt actually diverge instead of all
// pointing at one shared path-hash slug.
//
// Detached HEAD (empty branch) gets `main_detached_<hex>` keyed off
// the repo path so two detached repos don't collide. Empty branch is
// the edge case at boot before the daemon has detected HEAD.
func ForMain(worktreePath string, branch string) Slug {
	if branch == "" {
		sum := blake3.Sum256([]byte(worktreePath))
		return Slug{Value: "main_detached_" + hexEncode(sum[:])[:8], Source: SourceMain}
	}
	sanitised := sanitiseBranch(branch)
	if sanitised == "" {
		// Branch was symbol-only (`@`, `###`, etc — git allows it).
		// Falling through with an empty sanitised string would produce
		// `main_` which collides for every odd branch in this repo;
		// hash the raw branch instead so each one gets a stable slug.
		sum := blake3.Sum256([]byte(branch))
		return Slug{Value: "main_sym_" + hexEncode(sum[:])[:8], Source: SourceMain}
	}
	v := "main_" + sanitised
	if len(v) > 32 {
		// Same naive-truncation trap as extractTicket: collide-prone
		// without a hash suffix when the branch name is long.
		sum := blake3.Sum256([]byte(v))
		tag := hexEncode(sum[:])[:6]
		v = v[:32-1-len(tag)] + "_" + tag
	}
	return Slug{Value: v, Source: SourceMain}
}

// sanitiseBranch lowercases the branch name and replaces every
// non-alphanumeric run with a single underscore so the result is a
// valid identifier across MySQL/Postgres/Redis prefixes. Leading and
// trailing underscores are trimmed.
func sanitiseBranch(branch string) string {
	var b strings.Builder
	prevUnderscore := true
	for _, r := range strings.ToLower(branch) {
		switch {
		case r >= 'a' && r <= 'z', r >= '0' && r <= '9':
			b.WriteRune(r)
			prevUnderscore = false
		default:
			if !prevUnderscore {
				b.WriteByte('_')
				prevUnderscore = true
			}
		}
	}
	return strings.Trim(b.String(), "_")
}

// extractTicket pulls a Jira-style `[A-Z]+-\d+` ticket out of s and
// renders it `prefix_num` (lowercased). Length budgeting and the
// path-hash suffix are composeTicket's job, so no truncation or hashing
// happens here.
func extractTicket(s string) (string, bool) {
	caps := ticketRe.FindStringSubmatch(s)
	if caps == nil {
		return "", false
	}
	return strings.ToLower(caps[1]) + "_" + caps[2], true
}

// sysvCksum reproduces the POSIX `cksum` (SysV) algorithm. CRC-32
// over (data || length octets little-endian), polynomial
// 0x04C11DB7, initial 0, final NOT.
//
// We need this in-process because the bash hooks use POSIX `cksum`
// and treeman's redis indices must agree with hand-computed ones.
func sysvCksum(input []byte) uint32 {
	var crc uint32
	for _, b := range input {
		crc = crcUpdate(crc, b)
	}
	length := uint64(len(input))
	for length != 0 {
		crc = crcUpdate(crc, byte(length&0xff))
		length >>= 8
	}
	return ^crc
}

func crcUpdate(crc uint32, b byte) uint32 {
	crc ^= uint32(b) << 24
	for range 8 {
		if crc&0x80000000 != 0 {
			crc = (crc << 1) ^ 0x04C11DB7
		} else {
			crc <<= 1
		}
	}
	return crc
}

const hexDigits = "0123456789abcdef"

func hexEncode(b []byte) string {
	out := make([]byte, 0, len(b)*2)
	for _, x := range b {
		out = append(out, hexDigits[x>>4], hexDigits[x&0x0f])
	}
	return string(out)
}
